### PR TITLE
Update ASTC version

### PR DESCRIPTION
**Pull request by jherico**
## Description

* Update the astcenc library version to the most recent release.  
  * Detect SIMD support and build the corresponding atscenc variant
  * Fixes a number of potential memory leaks caused by the non-thread-safe code in the previous version of astcenc
* Update the mipmap generation code to allocate the required memory for all mips at once instead of mip by mip (less fragmentation, and should be faster)

## Performance

Ran the command_buffer_usage sample 5 times against this branch and against the current main branch, recording the _Time spent loading images_ log line for each run.  The target machine is a Ryzen Threadripper 1950X with 32 logical processor and for which the AVX2 ISA was detected and used.

| branch | average load time | standard deviation |
| - | - | - |
| main | 5.6638216 | 0.07 |
| astc | 3.3971104 | 0.06 |

This represents a 40% increase in ASTC image decompression performance.

Note that the only caveat for this is that the modules being used to detect the SIMD instruction set were only found under the license embedded in the added CMake files, which is a slight variant of the BSD-3 license.  Hopefully this shouldn't be a problem.

## General Checklist:

Please ensure the following points are checked:

- [X] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [X] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [X] I have commented any added functions (in line with Doxygen)
- [X] I have commented any code that could be hard to understand
- [X] My changes do not add any new compiler warnings
  - Strictly speaking there may be new or different compiler warnings inside the ASTC build, but that seems out of scope.
- [ ] My changes do not add any new validation layer errors or warnings
- [X] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [X] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [X] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [X] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

### CODE REVIEW

File: framework/scene_graph/components/image.cpp [hunk 1]
```diff
@@ -168,8 +168,7 @@ const std::vector<uint8_t> &Image::get_data() const
 
 void Image::clear_data()
 {
-	data.clear();
-	data.shrink_to_fit();
```
Comment (jherico) [inline framework/scene_graph/components/image.cpp]: `shrink_to_fit` doesn't guarantee deallocation.

**Comment by gary-sweet** [comment]
Is this PR also suffering from the CRLF issue? Whole files appear different.

File: framework/scene_graph/components/image.cpp [hunk 2]
```diff
@@ -1,370 +1,396 @@
...
+		next_height = std::max<uint32_t>(1u, next_height / 2);
+		next_size   = next_width * next_height * channels;
+		if (next_width == 1 && next_height == 1)
+		{
+			break;
+		}
+	}
+
+	return result;
+}
```
Comment (asuessenbach) [inline framework/scene_graph/components/image.cpp]: You're missing the 1x1 level.
Maybe better:
```
	constexpr uint32_t channels = 4;
	auto               width    = std::max<uint32_t>(1, extent.width);
	auto               height   = std::max<uint32_t>(1, extent.height);
	auto               size     = width * height * channels;
	auto               result   = size;
	while (size != channels)
	{
		width  = (width + 1) / 2;
		height = (height + 1) / 2;
		size   = width * height * channels;
		result += size;
	}

	return result;
```

File: framework/scene_graph/components/image.cpp [hunk 3]
```diff
@@ -1,370 +1,396 @@
...
+	}
+
+	auto extent      = get_extent();
+	auto next_width  = std::max<uint32_t>(1u, extent.width / 2);
+	auto next_height = std::max<uint32_t>(1u, extent.height / 2);
+	auto channels    = 4;
+	auto next_size   = next_width * next_height * channels;
+
+	// Allocate for all the mips at once
+	data.reserve(get_required_mipmaps_size(extent));
```
Comment (asuessenbach) [inline framework/scene_graph/components/image.cpp]: You probably wanted to say
`	data.reserve(data.size() + get_required_mipmaps_size(extent));
`

File: framework/scene_graph/components/image.cpp (hunk 2 above)
Comment (jherico) [reply framework/scene_graph/components/image.cpp]: Good point, but th suggested replacement fails as well for the degenerate case of a 1x1 image, so the while should be `do {...} while` I think.

File: framework/scene_graph/components/image.cpp (hunk 3 above)
Comment (jherico) [reply framework/scene_graph/components/image.cpp]: No, the mip size calculations above include lip #0.  I'll add a comment clarifying.

File: framework/scene_graph/components/image.cpp (hunk 2 above)
Comment (asuessenbach) [reply framework/scene_graph/components/image.cpp]: Ahm... no, I don't think it would fail.
`result` is initialized with the top-level `size`. If that top-level size is 1x1, it's correct.

Comment (jherico) [reply framework/scene_graph/components/image.cpp]: Sorry, you're right, I missed that.

**Review by gary-sweet — DISMISSED** [review]

**Review by tomadamatkinson — DISMISSED** [review]
Good improvement. Thanks for performance testing!

**Comment by SaschaWillems** [comment]
> Note that the only caveat for this is that the modules being used to detect the SIMD instruction set were only found under the license embedded in the added CMake files, which is a slight variant of the BSD-3 license. Hopefully this shouldn't be a problem.

We probably need to check this internally. @marty-johnson59: Maybe something to put on next calls agenda? Or forward to Legal at Khronos. It's a different license, so we need to make sure that we can use it and that we don't block any current user from using our samples by that new license.

**Comment by SaschaWillems** [comment]
Tough tbh. I'd rather prefer moving all the assets that use astc to basisu instead. This is a nice stopgap, but probably becomes obsolete once we update the assets.

File: framework/CMakeLists.txt [hunk 4]
```diff
@@ -532,6 +532,10 @@ if(${NEED_LINK_ATOMIC})
     target_link_libraries(${PROJECT_NAME} PUBLIC atomic)
 endif()
 
+# I'm uncertain why this has to be done explicitly, when the target_link_libraries call above should do it
```
Comment (SaschaWillems) [inline framework/CMakeLists.txt]: Is this comment still current? If so we should tackle the issue before merging.

File: third_party/CMakeLists.txt [hunk 5]
```diff
@@ -290,44 +290,61 @@ endif()
 add_library(stb INTERFACE)
 target_include_directories(stb INTERFACE "${CMAKE_CURRENT_SOURCE_DIR}/stb")
 
+
+# On an x86_64 macbook, AVX2 is selected, but then I get an error during the final link
```
Comment (SaschaWillems) [inline third_party/CMakeLists.txt]: Same as above. We either need to tackle this or reword the comment. Esp the "I..." part.

**Review by SaschaWillems — CHANGES_REQUESTED** [review]
LGTM aside from the legal stuff. I only have some minor remarks regarding some comments.

File: framework/CMakeLists.txt (hunk 4 above)
Comment (jherico) [reply framework/CMakeLists.txt]: It is.  When I removed that I saw compile errors where `astcenc.h` could not be found.  In theory, the `target_link_libraries` call with the astc target _should_ automatically add the interface include paths.

Comment (SaschaWillems) [reply framework/CMakeLists.txt]: So if it's not possible to fix this, could you rephrase the comment to not refer to you personally (the "I'm..." part). (can't explain it better, not a native English speaker).

Comment (jherico) [reply framework/CMakeLists.txt]: it must have been some transient glitch.  I can't reproduce the error now.

File: third_party/CMakeLists.txt (hunk 5 above)
Comment (jherico) [reply third_party/CMakeLists.txt]: fixed

**Review by tomadamatkinson — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]
Code LGTM.

We still need to check on licenses before merging.

**Comment by tomadamatkinson** [comment]
@marty-johnson59 is a licence check needed on this?

**Comment by marty-johnson59** [comment]
ATSC is under Apache 2 license, so we're OK with that.  Merging with 3..

**Comment by SaschaWillems** [comment]
Looks like this broke builds on recent Apple devices (e.g. M1), see https://github.com/KhronosGroup/Vulkan-Samples/issues/965

**Comment by kanavin** [comment]
@jherico @SaschaWillems there's another issue here: the newly added cmake modules check whether the compiler can produce the instructions, by trying various command line flags, and then forcibly enable them if so. 

This doesn't work when the build is intended for a machine that doesn't actually support those instructions; the checks should either try only the default compiler settings (e.g. no additional options), or have a switch to forcibly override ASTC_ARCH from cmake command line.

This was exposed by trying to build for 32 bit x86 (yes, it's still a valid architecture in 2024) - the checks happily enable AVX2 because 32bit x86 compiler can do it if you ask, even though the overall build is configured for a mid 2000s x86 CPU model that can't possibly have that.

**Comment by jherico** [comment]
@kanavin is it critical that this be automatic, or would it be sufficient if you could force the ISA from the command line, like for instance `cmake -Bbuild -DASTC_ARCH=NONE`?  

If the latter is sufficient, then I have a PR up that should allow you to build: #978

**Comment by kanavin** [comment]
> @kanavin is it critical that this be automatic, or would it be sufficient if you could force the ISA from the command line, like for instance `cmake -Bbuild -DASTC_ARCH=NONE`?
> 
> If the latter is sufficient, then I have a PR up that should allow you to build: #978

Something like this should work, yes. I've tested a local patch that does something similar, and was going to open a pull request:

https://github.com/kanavin/Vulkan-Samples/commit/41090ff333b2b4d888a15e3fba000229fc8740c3

While you're at it, SSE41 should become sse4.1 (with a dot), when lowercased, as that's how astc defines the target.

**Comment by jherico** [comment]
Ugh, you're right... 🤦

image

I'm just going to exclude it for now.  It's a relatively narrow window of processors that support SSE 4.1 and NOT AVX2.
